File: src/main/scala/darkyenus/lowscape/world/terrain/TerrainPatchwork.scala
package darkyenus.lowscape.world.terrain

import com.badlogic.gdx.graphics.Camera
import com.badlogic.gdx.graphics.g3d.{Renderable, RenderableProvider, Shader}
import com.badlogic.gdx.math.collision.BoundingBox
import com.badlogic.gdx.math.{MathUtils, Vector3}
import com.badlogic.gdx.utils.{Array => GArray, Pool}

/**
 * Private property.
 * User: Darkyen
 * Date: 08/12/14
 * Time: 20:13
 */
final class TerrainPatchwork(val patchAmount:Int,val patchSize:Int, val camera:Camera, patchMinHeight:Float = -100f, patchMaxHeight:Float = 1000f) extends RenderableProvider {

  val heights = Array.ofDim[Float](patchAmount * patchSize - patchAmount + 1,patchAmount * patchSize - patchAmount + 1)

  val terrainShader:Shader = null

  val patches = Array.tabulate[TerrainPatch](patchAmount * patchAmount)((i) => {
    val x = i % patchAmount
    val y = i / patchAmount
    val boundingBox = new BoundingBox(
      new Vector3(x * patchSize, y * patchSize, patchMinHeight),
      new Vector3(x * patchSize + patchSize, y * patchSize + patchSize, patchMaxHeight))

    val patch = new TerrainPatch(heights,patchSize,x * patchSize - x,y * patchSize - y, boundingBox = boundingBox, shader = terrainShader)
    patch.updateRenderables()
    patch
  })

  def generateMesh(tabulator:(Int,Int) => Float): Unit ={
    var x = 0
    var y = 0
    while(x < heights.length){
      while(y < heights.length){
        heights(x)(y) = tabulator(x,y)
        y += 1
      }
      y = 0
      x += 1
    }
  }

  def updateMesh(): Unit ={
    var i = 0
    while(i < patches.length){
      patches(i).updateMesh()
      i += 1
    }
  }

  updateMesh()

  override def getRenderables(renderables: GArray[Renderable], pool: Pool[Renderable]): Unit = {
    var i = 0
    val frustum = camera.frustum
    while(i < patches.length){
      val patch = patches(i)
      if(frustum.boundsInFrustum(patch.boundingBox)){
        patch.getRenderables(renderables,pool)
      }
      i += 1
    }
  }

  private val worldSizeLimit = patchAmount * patchSize - patchAmount
  def heightAt(x:Float,y:Float):Float = {
    val lowX = x.toInt
    val lowY = y.toInt
    if(lowX < 0 || lowY < 0 || lowX >= worldSizeLimit || lowY >= worldSizeLimit){
      return 0
    }
    val bottomX = MathUtils.lerp(heights(lowX)(lowY),heights(lowX + 1)(lowY),x - lowX)
    val topX = MathUtils.lerp(heights(lowX)(lowY + 1),heights(lowX + 1)(lowY+1),x - lowX)
    MathUtils.lerp(bottomX,topX,y - lowY)
  }
}

File: src/main/scala/darkyenus/lowscape/world/terrain/TerrainPatchwork.java
package darkyenus.lowscape.world.terrain;

import com.badlogic.gdx.graphics.Camera;
import com.badlogic.gdx.graphics.g3d.Renderable;
import com.badlogic.gdx.graphics.g3d.RenderableProvider;
import com.badlogic.gdx.graphics.g3d.Shader;
import com.badlogic.gdx.math.Frustum;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.math.collision.BoundingBox;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.Pool;

/**
 * @author Darkyen
 */
public class TerrainPatchwork implements RenderableProvider {

    private final Camera camera;
    private final float[][] heights;
    private final TerrainPatch[] patches;
    private final int worldSizeLimit;

    public TerrainPatchwork(int patchAmount, int patchSize, Camera camera) {
        this.camera = camera;

        final float patchMinHeight = -100f, patchMaxHeight = 1000f;
        final Shader terrainShader = null;

        heights = new float[patchAmount * patchSize - patchAmount + 1][patchAmount * patchSize - patchAmount + 1];

        patches = new TerrainPatch[patchAmount * patchAmount];
        for (int i = 0; i < patches.length; i++) {
            final int x = i % patchAmount;
            final int y = i / patchAmount;
            final BoundingBox boundingBox = new BoundingBox(
                    new Vector3(x * patchSize, y * patchSize, patchMinHeight),
                    new Vector3(x * patchSize + patchSize, y * patchSize + patchSize, patchMaxHeight));

            TerrainPatch patch = new TerrainPatch(heights,patchSize,x * patchSize - x,y * patchSize - y, boundingBox, terrainShader);
            patch.updateRenderables();
            patches[i] = patch;
        }

        worldSizeLimit = patchAmount * patchSize - patchAmount;

        updateMesh();
    }

    public void generateMesh(TerrainGenerator generator){
        for (int x = 0, h = heights.length; x < h; x++) {
            for (int y = 0; y < h; y++) {
                heights[x][y] = generator.getHeight(x,y);
            }
        }
    }

    @FunctionalInterface
    public interface TerrainGenerator {
        float getHeight(int x,int y);
    }

    public void updateMesh(){
        for(TerrainPatch patch:patches){
            patch.updateMesh();
        }
    }

    public final float heightAt(float x, float y) {
        final int lowX = (int) x;
        final int lowY = (int) y;
        if(lowX < 0 || lowY < 0 || lowX >= worldSizeLimit || lowY >= worldSizeLimit){
            return 0;
        }
        final float bottomX = MathUtils.lerp(heights[lowX][lowY],heights[lowX + 1][lowY],x - lowX);
        final float topX = MathUtils.lerp(heights[lowX][lowY + 1],heights[lowX + 1][lowY+1],x - lowX);
        return MathUtils.lerp(bottomX, topX, y - lowY);
    }

    @Override
    public void getRenderables(Array<Renderable> renderables, Pool<Renderable> pool) {
        final Frustum frustum = camera.frustum;
        for(TerrainPatch patch: patches){
            if(frustum.boundsInFrustum(patch.boundingBox)){
                patch.getRenderables(renderables,pool);
            }
        }
    }
}
